### PR TITLE
Java requirement

### PR DESCRIPTION
**Pull request by coreone**
I found when installing zookeeper that some systems did not have Java installed already, so I added the ability to set a flag to require a Java package be installed, along with a variable to set the name of the Java package.  This was only done for RHEL systems so far since I don't have access to a Debian one right now to test on, but it should be easy enough to potentially just copy the same code from redhat.pp to debian.pp to get it working.

### CODE REVIEW

**Comment by deric** [comment]
ZooKeeper package should have dependency on Java (in case of Debian `default-jdk` or something like this is fine).  I'm not sure if Cloudera packages respect this.

Anyway, to make this feature working, we have to ensure that Java will be installed before ZooKeeper package. A cleanest solution probably is using an `Anchor` - install Java before this anchor and ZooKeeper packages afterwards.

File: manifests/install.pp [hunk 1]
```diff
@@ -38,6 +40,8 @@
         packages          => $packages,
         before            => Anchor['zookeeper::install::end'],
         require           => Anchor['zookeeper::install::begin'],
+        install_java      => $install_java,
+        java_package      => $java_package
```
Comment (deric) [inline manifests/install.pp]: this should be in `RedHat` case

**Comment by coreone** [comment]
Fix that dumb parameter location issue, but I'm still working on the anchor piece.  I haven't done anything with anchors in Puppet yet, so I'm still looking into how to best do this.  I would imagine wrapping the area around the Java installation in redhat.pp with anchors like "zookeeper::install::java::begin" and "zookeeper::install::java::end" would work, but I'm trying to figure out how to get ensure_package to use those anchors...

**Comment by deric** [comment]
It's quite easy, just define an anchor:

```puppet
anchor { 'zookeeper::install::java': }
```
and then we want `(install java package) -> (install zookeeper package)`. So we can use `before` hook:

```puppet
package{ $java_package:
    ensure => present,
    allow_virtual => true,
    before  => Anchor['zookeeper::install::java']
}
```
and for ZooKeeper we need "after" hook, which is equivalent to:

```puppet
    require  => Anchor['zookeeper::install::java']
```
The code above could be used with a class, however with function `ensure_resource` it's a bit complicated. Maybe something like:

```puppet
Anchor['zookeeper::install::java']
->
ensure_resource('package', $packages, {'ensure' => $ensure})
```
Sorry, I haven't tried to compile it, but it might work :)

**Comment by coreone** [comment]
I tried something a little different.  I changed the ensure_resource to just a package declaration, and then surrounded that by anchors.  I then put a "before" requirement for the first anchor in the java package resource.  It seemed to work in my testing, and did indeed install Java first.

**Comment by deric** [comment]
This code 
```puppet
anchor { 'zookeeper::install::package::begin': }->
  package { $packages:
    ensure => present
}->
anchor { 'zookeeper::install::package::end': }
```
should be the same as:
```puppet
package { $packages:
    ensure => present,
    before => Anchor['zookeeper::install::package::begin'],
    require => Anchor['zookeeper::install::package::end'] 
}
```
Difference between declaring directly `package {'zookeeper': ... }` and using `ensure_resource(...)` is that the latter does not throw exception when `zookeeper` was declared elsewhere (in Puppet code). Declaring ZooKeeper just in one place is quite all right (it is good practice, at least). However with Java it's something completely different. Some people might install Oracle Java, other openJDK and moreover Java will be declared in many Puppet modules. In order to make this feature useful, Java has to be declared with `ensure_resource`.

Btw. now you broke the possibility to uninstall ZooKeeper.

**Comment by coreone** [comment]
I tried another pass at this and it doesn't seem to work correctly either.  I also talked to my coworker who runs our Puppet service here, and he suggested that I just use a Java module to require the Java package on the system instead of trying to build a Java installation into a separate module.  So, do you think we should just abandon this and not try to do a Java installation inside the Zookeeper module?  I'm up for just using an outside module if that makes more sense to you, so feel free to delete this pull request if that sounds reasonable.

Thanks for all the anchor help, though!

File: manifests/os/redhat.pp [hunk 2]
```diff
@@ -8,12 +8,31 @@
...
+  if ($install_java){
+    if !$java_package {
+      fail { "Java installation is required, but no java package was provided.": }
+    }
+
+    validate_string($java_package)
+
+    # make sure the Java package is only installed once.
+    anchor { 'zookeeper::install::package::begin': }
+    ensure_resource('package', $java_package, {'ensure' => $ensure, 'allow_virtual' => true})
```
Comment (deric) [inline manifests/os/redhat.pp]: There's no guarantee that this line will be executed "between" those anchors. Otherwise it looks good, well done :+1:

**Pull request merged**